sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/CancellationTokenSourceExtensions.cs

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+
+namespace Azure.Messaging.ServiceBus
+{
+    internal static class CancellationTokenSourceExtensions
+    {
+        public static void CancelAfterLockExpired(this CancellationTokenSource cancellationTokenSource,
+            ServiceBusReceivedMessage receivedMessage)
+        {
+            if (cancellationTokenSource == null || receivedMessage.LockedUntil == default)
+            {
+                return;
+            }
+
+            var delay = receivedMessage.LockedUntil.Subtract(DateTimeOffset.UtcNow);
+            if (delay > TimeSpan.Zero)
+            {
+                cancellationTokenSource.CancelAfter(delay);
+            }
+            else
+            {
+                cancellationTokenSource.Cancel();
```

Comment:
I don't think we want to cancel the token ourselves based on the expiry time. The expiry time can change in the background as we renew the lock, and the service does not always honor the expiry time exactly - for instance, for session locks it is more lenient if you are actively receiving from the session. I think we should restrict this feature to cases where we know the lock has been lost, namely when we handle ServiceBusException with a Reason of SessionLockLost.

Reply:
>  The expiry time can change in the background as we renew the lock

That case would be covered because when the lock is renewed the cancellation token source gets prolonged by calling `CancelAfter` again which moves the cancellation forward. 

https://github.com/Azure/azure-sdk-for-net/pull/36253/files#diff-80a871db0c5d5f587285b6b70833d8a0a70b3abde8484e009eca4d5a67434896R311

Given that behavior would it be ok to leave this in and address

> and the service does not always honor the expiry time exactly - for instance, for session locks it is more lenient if you are actively receiving from the session

by checking for session receivers not then on those cases not cancel based on the lock until?

Reply:
FYI I already asked for some clarification a while ago in https://github.com/Azure/azure-sdk-for-net/issues/32271#issuecomment-1363776639 because I didn't fully get your previous reasoning either.

Reply:
I'm suggesting that we only cancel the LockLostCancellationToken when we get back a SessionLockLost from the service.

Reply:
I was thinking I begin to understand that we should never do `CancelAfter` based on the LockExpiry time but the comment 

> Also, the user can always check the Lock Expiry time themselves if they want to use this. 

confused me because if the user can check the property and it has "meaning" why would it not have meaning for the lib itself.

Reply:
The user is free to take any action that they wish based on the Lock Expiry Time and their application scenario. To me, this is different than the library saying that after this expiry time has passed, the message lock has definitely been lost. The value I see in this feature (and I believe this matches what users have reported) is to give users a way to proactively end processing when the library knows definitively that the lock is lost, based on the service response.

Reply:
> I'm suggesting that we only cancel the LockLostCancellationToken when we get back a MessageLockLost from the service.

Isn't then the value of this feature greatly reduced? We can only observe the LockLost for messages that get auto renewed. It could also be confusing for users because as far as I understand the LockUntil might have a value that is expired but the token doesn't get cancelled. But I guess we are trying to make sure we can complete as long as the service allows to complete because the final settlement call or the renewal is the only true "source of truth". Is my understanding correct?

Reply:
> But I guess we are trying to make sure we can complete as long as the service allows to complete because the final settlement call or the renewal is the only true "source of truth". Is my understanding correct?

Yes, that's exactly right. I would rather err on the side of not saying the lock is lost vs saying the lock is lost.

Reply:
I think that reasoning is sound. Happy to change it.

Reply:
> Isn't then the value of this feature greatly reduced? We can only observe the LockLost for messages that get auto renewed. It could also be confusing for users because as far as I understand the LockUntil might have a value that is expired but the token doesn't get cancelled.

Yeah that is a fair point - I will think about it a bit more. Since this is an opt in feature for users, perhaps we can document the caveat that the token may be signaled in some cases when it would still be possible to settle the message.

Reply:
I will wait with the changes a few more days then 😉

Reply:
There's an inherent potential for being inaccurate due to clock skew - expiring early or late - that I think is the same concern as the service not strictly enforcing the time.  

Josh is correct in that the only sure way to know that the lock has expired is to catch the error condition.   However, I do agree that there's merit to the approach as it gives me a simpler way as an author to react to "there's a risk of the token being expired" than having to know what part of the message to observe.  It at least raises my awareness of the risk and lets me proactively decide not to invoke costly or lengthy processing.

I'm in favor of the current approach, so long as the feature is an opt-in for timing-based cancellation.

Reply:
Yep I mentioned the clock skew already in a previous comment and it should also be clear from the XML documentation. Any suggestions how to call this feature flag on the processor options? Happy to add that to the PR.

Reply:
So we discussed this offline and we are in agreement with your proposed changes that leverage the expiry time! Would you be willing to add a similar lock token for the Session case in ProcessSessionMessageEventArgs?

Reply:
Sure can do

Reply:
@JoshLove-msft I was trying to wrap my head around the session part and as a first go I did use the same approach already available in the receive actions to tie the token source to the `LockedUntil`. When the user calls `RenewSessionLock` on the session processor args I can easily prolong the cancellation token by using the `sessionReceiver.SessionLockedUntil`. But for the auto-renewal I have no link anymore to the currently received messages and have to chance to refresh the token sources.

Is my understanding correct that the expiry time of a message is aligned with the SessionLockUntil? If that is the case I wonder how much value the additional token would provide given the session processor arg cancellation token is already wired to session lock lost.

To make it truly valuable in the session auto-renewal case I would have to keep a weak reference around in the session receive manager to the received messages and the refresh the LockUntil when the session lock renewal was successful or cancel the token source when it wasn't but that might be tricky to get right. 

Any thoughts?

Reply:
Any thoughts on the above @JoshLove-msft @jsquire ?

Reply:
I'm going to defer to @JoshLove-msft, but the general structure looks reasonable to me.  Personally, I'd bypass the .NET events and just assign Func instances since this is an internal implementation and not part of the public API.

Reply:
I pushed the approach. Let's see what @JoshLove-msft thinks.

Reply:
> But for the auto-renewal I have no link anymore to the currently received messages and have to chance to refresh the token sources.

I may be missing your point, but I don't think we need the link to the received messages. The cancellation token should be per-session for the session case. It would be similar to the existing CancellationToken property with the additional logic to make it get signaled based on the session receiver SessionLockedUntil property.

Reply:
Well I'm figuring things out while we go. I haven't done much with sessions and you asked originally "can you do the same for the sessions" which cause me to think / assume the relationship is also related to messages

Reply:
Here is the LinkedTokenSource creation https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs#L293

Reply:
I still think the CancellationToken and SessionLockCancellationToken logic will be different for sessions:
CancellationToken - signaled when processor stopping or session lock lost exception
SessionLockCancellationToken - signaled when SessionLockedUntil time passes or session lock lost exception